src/colmap/estimators/manifold.h

```diff
@@ -0,0 +1,156 @@
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <ceres/ceres.h>
+#include <ceres/rotation.h>
+
+namespace colmap {
+
+#include <cmath>
+
+#include "ceres/ceres.h"
+
+inline void SetQuaternionManifold(ceres::Problem* problem, double* quat_xyzw) {
+#if CERES_VERSION_MAJOR >= 3 || \
+    (CERES_VERSION_MAJOR == 2 && CERES_VERSION_MINOR >= 1)
+  problem->SetManifold(quat_xyzw, new ceres::EigenQuaternionManifold);
+#else
+  problem->SetParameterization(quat_xyzw,
+                               new ceres::EigenQuaternionParameterization);
+#endif
+}
+
+inline void SetSubsetManifold(int size,
+                              const std::vector<int>& constant_params,
+                              ceres::Problem* problem,
+                              double* params) {
+#if CERES_VERSION_MAJOR >= 3 || \
+    (CERES_VERSION_MAJOR == 2 && CERES_VERSION_MINOR >= 1)
+  problem->SetManifold(params,
+                       new ceres::SubsetManifold(size, constant_params));
+#else
+  problem->SetParameterization(
+      params, new ceres::SubsetParameterization(size, constant_params));
+#endif
+}
+
+template <int size>
+inline void SetSphereManifold(ceres::Problem* problem, double* params) {
+#if CERES_VERSION_MAJOR >= 3 || \
+    (CERES_VERSION_MAJOR == 2 && CERES_VERSION_MINOR >= 1)
+  problem->SetManifold(params, new ceres::SphereManifold<size>);
+#else
+  problem->SetParameterization(
+      params, new ceres::HomogeneousVectorParameterization(size));
+#endif
+}
+
+// Use an exponential function to ensure the variable to be strictly positive
+// Generally for the scale parameters
+#if CERES_VERSION_MAJOR >= 3 || \
+    (CERES_VERSION_MAJOR == 2 && CERES_VERSION_MINOR >= 1)
+template <int size = 1>
+class PositiveExponentialManifold : public ceres::Manifold {
+ public:
+  PositiveExponentialManifold() {}
+  ~PositiveExponentialManifold() {}
+
+  bool Plus(const double* x,
+            const double* delta,
+            double* x_plus_delta) const override {
+    for (size_t i = 0; i < size; ++i) {
+      x_plus_delta[i] = std::exp(std::log(x[i]) + delta[i]);
```

Comment:
Isn't this equivalent to `x[i] * std::exp(delta[i])` ?

Reply:
Yes it is. Let me update.